internal/core/app_deploy.go

```diff
@@ -150,11 +151,16 @@ func (op *deployOperation) Upsert(
 	return resp.Deployment, nil
 }
 
-func (op *deployOperation) Do(ctx context.Context, log hclog.Logger, app *App, _ proto.Message) (interface{}, error) {
+func (op *deployOperation) Do(ctx context.Context, log hclog.Logger, app *App, msg proto.Message) (interface{}, error) {
 	// Sync our config first
 	if err := app.ConfigSync(ctx); err != nil {
 		return nil, err
 	}
+	// Upsert immediately so that if deployment fails we still have a reference to it in the db
```

Comment:
This is a stop-gap measure. I suspect the larger plan is to track all pipeline lifecycle events in the state, but this had the effect, in this case, of making a hung deployment more obvious to the user. In my case, I was experiencing a timeout due to a crash-loop backoff. So this was nice to have. Maybe the long term vision is something like an FSM. Not sure.

Reply:
Indeed, I have to think about this more but this change makes me feel slightly uncomfortable for the reasons you stated.